pkg/filters/fields.go

```diff
@@ -81,15 +81,48 @@ type FieldFilter struct {
 	fields         fmutils.NestedMask
 	action         tetragon.FieldFilterAction
 	invertEventSet bool
+	needsCopy      map[tetragon.EventType]struct{}
 }
 
 // NewFieldFilter constructs a new FieldFilter from a set of fields.
 func NewFieldFilter(eventSet []tetragon.EventType, fields []string, action tetragon.FieldFilterAction, invertEventSet bool) *FieldFilter {
+	// We only need to copy exec and exit events when they are explicitly filtering out
+	// the PID. This is because we need the PID to not be nil when accessing the event
+	// later on from the eventcache. See additional comments below for details.
+	var maybeNeedsCopy bool
+	if action == tetragon.FieldFilterAction_EXCLUDE {
+		for _, field := range fields {
+			if strings.HasPrefix(field, "process") {
+				maybeNeedsCopy = true
+				break
+			}
+		}
+	} else if action == tetragon.FieldFilterAction_INCLUDE {
+		// For inclusion, it's the opposite situation from the above. If the process.pid
+		// field is NOT present, it will be trimmed. So assume we need a copy unless we
+		// see process.pid.
+		maybeNeedsCopy = true
+		for _, field := range fields {
+			if field == "process.pid" {
```

Comment:
Does this also need HasPrefix()?

Reply:
It has to be an exact match so that it's more conservative. If we used HasPrefix and you did something like `process.binary` then we would not do a copy but PID would still be missing.

Reply:
Hopefully that makes sense